api/python/quilt3/packages.py

```diff
@@ -880,10 +914,49 @@ def set(self, logical_key, entry=None, meta=None):
             entry = PackageEntry([url], size, None, orig_meta)
         elif isinstance(entry, PackageEntry):
             entry = entry._clone()
+
+        elif object_is_serializable(entry):
+            # Use file extension from serialization_location, fall back to file extension from logical_key
+            ext = None
+            if len(pathlib.Path(logical_key).suffix) > 0:
+                ext = pathlib.Path(logical_key).suffix[1:]
+            if serialization_location:
+                if len(pathlib.Path(serialization_location).suffix) > 0:
+                    ext = pathlib.Path(serialization_location).suffix[1:]
+
+            format_handlers = FormatRegistry.search(type(entry), ext=ext)
+            if ext:
+                format_handlers = [f for f in format_handlers if ext in f.handled_extensions]
+
+            if len(format_handlers) == 0:
+                error_message = f'Quilt does not know how to serialize a {type(entry)}'
+                error_message_fragment = "."
+                if ext is not None:
+                    error_message_fragment = f' as a {ext} file.'
+                error_message += error_message_fragment + " "
+                error_message += f'If you think this should be supported, please open an issue or PR at ' \
+                                 f'https://github.com/quiltdata/quilt'
+                raise QuiltException(error_message)
+
+            if serialization_format_opts is None:
+                serialization_format_opts = {}
+            serialized_object_bytes, new_meta = format_handlers[0].serialize(entry, meta=None, ext=ext,
+                                                                             **serialization_format_opts)
+            if serialization_location:
+                serialization_path = pathlib.Path(serialization_location).expanduser().absolute()
+                serialization_path.parent.mkdir(exist_ok=True, parents=True)
+            else:
+                suffix = f'.{ext}' if ext else None
+                tmpfile = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
```

Comment:
Suggestion (and feel free to implement in a separate PR):
Serialized object files should go somewhere in the local package registry (i.e. under APP_DIR). That way all the quilt3 state will be in one place and a lot easier to track down. Also, temp files can get deleted even if the user saves the package manifest.

Reply:
> Serialized object files should go somewhere in the local package registry (i.e. under APP_DIR)

Are you thinking in the case that the user doesn't specify a serialization location or in general?

> Also, temp files can get deleted even if the user saves the package manifest.

It's always possible that someone removes the underlying files on disk so won't this be true no matter where we write to?

Reply:
TemporaryFiles are deleted automatically (at least by default I think). Storing serialization files in the local Quilt registry seems a lot safer to me.

Reply:
In this case the danger of bloating local disk is worse. I think TempFiles are appropriate.

Reply:
Adding it to my future todo. These tempfiles aren't actually temporary as far as I can tell (if delete=False, the file isn't cleaned up on file close and it writes to `/var/folder/...` which as far as I know isn't ever automatically cleaned up.